### ISSUE TITLE
react-onclickoutside does not work correctly for shadow DOM

### ISSUE THREAD

**Issue by dannysindra** [Hacker0x01/react-datepicker#2555]
**Describe the bug**
Bug report: in shadow DOM, any clicks inside the calendar will be treated as outside-clicks. In shadow DOM component, all `event.target` will be [re-targeted](https://javascript.info/shadow-dom-events) by browser to the shadow host. This means whenever end-user clicks on any part of the calendar (nav, day) the `handleCalendarClickOutside` callback will be triggered and closing the calendar immediately.

Looking at the previous report in #1928 and some open PRs on `react-onclickoutside` git repo,

- [285](https://github.com/Pomax/react-onclickoutside/issues/285)
- [323](https://github.com/Pomax/react-onclickoutside/pull/323)
- [324](https://github.com/Pomax/react-onclickoutside/pull/324)

It is clear that `react-onclickoutside` repo is no longer actively maintained (last commit was on Aug 2019). We need another solution to fix this issue for clients who use shadow DOM.

**To Reproduce**
Same steps as reported in #1928. It was closed by bot without any resolution.

**Expected behavior**
In shadow DOM env, any clicks inside the calendar should not be treated as outside-clicks.

**Additional context**
This is the main blocker for our team to adopt react-datepicker. I personally love it much compared to other libraries, but the capability to work on shadow DOM env is a requirement for us.

**Comment by gurschitz** [comment]
Hey, we are also running into this problem, where we would need to use `react-datepicker` within a shadow DOM instance. Since `react-onclickoutside` is not maintained anymore, maybe it makes sense to remove this dependency altogether and bring in the necessary code into this library for handling outside clicks? 
 
We could make a PR for this if this would be acceptable. :)

**Comment by fedoryakubovich** [comment]
Do you have any updates on that? Are there any ways to get around this?

**Comment by conoremclaughlin** [comment]
I'm also running into this issue. Possible for one of the team to fork react-onclickoutside and use one of the above PRs as the dependency?

**Comment by fedoryakubovich** [comment]
@conoremclaughlin Thanks a lot! It works for me. You saved a lot of time for me.

**Comment by mariyan-borisov** [comment]
@conoremclaughlin's changes have been merged upstream. His hacky way of modifying `yarn.lock` is no longer needed - just make sure to update `react-onclickoutside` to version ^6.12.0.

`react-datepicker` seems to work inside an *open* shadow root, but it still doesn't work inside a *closed* shadow root. I found [a hacky workaround](https://github.com/Pomax/react-onclickoutside/issues/285) by adding [`react-datepicker-ignore-onclickoutside`](https://github.com/Hacker0x01/react-datepicker/blob/0c07cb745c5af7679d9e85621c05d87ccc1b8e6e/src/index.jsx#L48) as a class to the shadow host. This seems to work only when the user clicks outside the shadow host. That's the best I got so far, and I'm not sure whether there's a *proper* way to fix it in `react-onclickoutside`.
